src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -2086,37 +2016,44 @@ private async Task CopyToContentLengthAsync(Stream destination, bool async, ulon
                     // if we filled the read buffer while doing the last read (which is at least one indication that the
                     // data arrival rate is fast enough to warrant a larger buffer), and if the buffer size we'd want is
                     // larger than the one we already have, then grow the connection's read buffer to that size.
-                    if (origReadBuffer == null)
+                    if (origReadBuffer is null)
                     {
-                        byte[] currentReadBuffer = _readBuffer;
-                        if (remaining == currentReadBuffer.Length)
+                        int currentCapacity = _readBuffer.Capacity;
+                        if (remaining == currentCapacity)
                         {
                             int desiredBufferSize = (int)Math.Min((ulong)bufferSize, length);
-                            if (desiredBufferSize > currentReadBuffer.Length)
+                            if (desiredBufferSize > currentCapacity)
                             {
-                                origReadBuffer = currentReadBuffer;
-                                _readBuffer = ArrayPool<byte>.Shared.Rent(desiredBufferSize);
+                                origReadBuffer = _readBuffer.DangerousGetUnderlyingBuffer();
+                                byte[] pooledBuffer = ArrayPool<byte>.Shared.Rent(desiredBufferSize);
+                                _readBuffer = new ArrayBuffer(pooledBuffer);
                             }
                         }
                     }
                 }
             }
             finally
             {
-                if (origReadBuffer != null)
+                if (origReadBuffer is not null)
                 {
-                    byte[] tmp = _readBuffer;
-                    _readBuffer = origReadBuffer;
-                    ArrayPool<byte>.Shared.Return(tmp);
+                    Debug.Assert(origReadBuffer.Length > 0);
 
                     // _readOffset and _readLength may not be within range of the original
```

Comment:
the mention of _readOffset and _readLength does not make sense anymore. Please check also other comments throughout the file

It migh talso be more clear if the comment was moved directly above the newly added `if` below as that is the part that would otherwise cause the most confusion

Reply:
Looks like these were the only remaining mentions of the old fields.